utils_cv/action_recognition/dataset.py

```diff
@@ -0,0 +1,473 @@
+        self.test_transforms = test_transforms
+        self.random_shift = random_shift
+        self.temporal_jitter = temporal_jitter
+        self.batch_size = batch_size
+        self.video_ext = video_ext
+        self.warning = warning
+
+        # create training and validation datasets
+        self.train_ds, self.test_ds = (
+            self.split_with_file(
+                train_split_file=train_split_file,
+                test_split_file=test_split_file,
+            )
+            if train_split_file
+            else self.split_train_test(train_pct=train_pct)
+        )
+
+        # initialize dataloaders
+        self.init_data_loaders()
+
+    def split_train_test(
+        self, train_pct: float = 0.8
+    ) -> Tuple[Dataset, Dataset]:
+        """ Split this dataset into a training and testing set
+
+        Args:
+            train_pct: the ratio of images to use for training vs
+            testing
+
+        Return
+            A training and testing dataset in that order
+        """
+        pass
+
+    def split_with_file(
+        self,
+        train_split_file: Union[Path, str],
+        test_split_file: Union[Path, str],
+    ) -> Tuple[Dataset, Dataset]:
+        """ Split this dataset into a training and testing set using a split file.
+
+        Args:
+            split_files: a tuple of 2 files
+
+        Return:
+            A training and testing dataset in that order
+        """
+        self.video_records = []
+
+        # add train records
+        self.video_records.extend(
+            [VideoRecord(x.strip().split(" ")) for x in open(train_split_file)]
+        )
+        train_len = len(self.video_records)
+
+        # add validation records
+        self.video_records.extend(
+            [VideoRecord(x.strip().split(" ")) for x in open(test_split_file)]
+        )
+
+        # create indices
+        indices = torch.arange(0, len(self.video_records))
+        train_range = indices[:train_len]
+        test_range = indices[train_len:]
+
+        # create train subset
+        train = copy.deepcopy(Subset(self, train_range))
+        train.dataset.transforms = self.train_transforms
+        train.dataset.sample_step = (
+            self.temporal_jitter_step
+            if self.temporal_jitter
+            else self.sample_step
+        )
+        train.dataset.presample_length = self.sample_length * self.sample_step
+
+        # create test subset
+        test = copy.deepcopy(Subset(self, test_range))
+        test.dataset.transforms = self.test_transforms
+        test.dataset.random_shift = False
+        test.dataset.temporal_jitter = False
+
+        return train, test
+
+    def init_data_loaders(self) -> None:
+        """ Create training and validation data loaders. """
+        devices = num_devices()
+
+        self.train_dl = DataLoader(
+            self.train_ds,
+            batch_size=self.batch_size * devices,
+            shuffle=True,
+            num_workers=0,
```

Comment:
Instead of num_workers=0, I guess we want to use: num_workers=db_num_workers(),

Reply:
Torch 1.2 has a bug when num-workers > 0 (0 means run a main-processor worker) - added this as a note

Reply:
Oh, is there any way around this? Training might be a lot slower with only 0 workers. And are you sure, we use >0 workers also for detection?

Reply:
My guess is that it will work with >0, ie setting 
   num_workers = db_num_workers()